index.js

```diff
@@ -531,11 +565,45 @@ class KeyringController extends EventEmitter {
         return { type, data };
       }),
     );
-    const encryptedString = await this.encryptor.encrypt(
-      this.password,
-      serializedKeyrings,
-    );
-    this.store.updateState({ vault: encryptedString });
+
+    let vault;
+    let newExportedKeyString;
+
+    if (this.cacheEncryptionKey) {
+      if (this.password) {
+        const { vault: newVault, exportedKeyString } =
+          await this.encryptor.encryptWithDetail(
+            this.password,
+            serializedKeyrings,
+          );
+
+        vault = newVault;
+        newExportedKeyString = exportedKeyString;
+      } else if (encryptionKey) {
+        const key = await this.encryptor.createKeyFromString(encryptionKey);
+        const vaultJSON = await this.encryptor.encryptWithKey(
+          key,
+          serializedKeyrings,
+        );
+        vaultJSON.salt = this.encryptionSalt;
+        vault = JSON.stringify(vaultJSON);
+      }
+    } else {
+      vault = await this.encryptor.encrypt(this.password, serializedKeyrings);
+    }
+
+    if (!vault) {
+      throw new Error('Cannot persist vault without vault information');
+    }
+
+    this.store.updateState({ vault });
+
+    // The keyring updates need to be announced before updating the exportedKeyStrings
```

Comment:
What is the relationship between these two updates? It's not clear to me why this order is important.

Reply:
Grrrr.  It's a weird extension issue.  If I call `this.memStore.updateState({ encryptionKey: newExportedKeyString });` *before* `_updateMemStoreKeyrings`, the account is created in the keyring *but* somehow isn't being propagated to the extension before the Preferences controller tries to `syncAddresses`.

Reply:
The visual side effect is that the new account wouldn't be selected in the extension -- Account 1 would be selected.  When you open up the accounts menu, however, it's there.

Reply:
This comment could be covering the why insteadof/alongwith the what

Reply:
Also, could a test cover this to guarantee it's not undone?

Reply:
> If I call `this.memStore.updateState({ encryptionKey: newExportedKeyString });` *before* `_updateMemStoreKeyrings`, the account is created in the keyring *but* somehow isn't being propagated to the extension before the Preferences controller tries to `syncAddresses`.  

Is this also why the calls to `_updateMemStoreKeyrings` were removed in a number of places? I was curious about that as well.

Reply:
Huh this is pretty tricky. I think I follow though.

Whenever `memStore` is updated, it'll trigger some other things to run (including `syncAddresses`). By setting the encryption key here before calling `_updateMemStoreKeyrings`, these state update get triggered with an _outdated_ set of keyrings and associated accounts.

An update with stale keyrings/accounts is a problem because it has side-effects like choosing a new selected address in some cases. Whereas a stale (i.e. unset) encryption key is not a problem at all. It just gets set in the next update.